flixel/graphics/frames/FlxAtlasFrames.hx

```diff
@@ -414,24 +412,158 @@ class FlxAtlasFrames extends FlxFramesCollection
 	}
 }
 
-typedef TexturePackerObject =
+typedef AtlasBase<T> = { frames:T }
+
+typedef AseAtlasBase<T> = AtlasBase<T> &
 {
-	frames:Dynamic
+	var meta:AseAtlasMeta;
+}
+
+typedef AseAtlasArray = AseAtlasBase<Array<AseAtlasFrame>>;
+typedef TexturePackerAtlasArray = AtlasBase<Array<AseAtlasFrame>>;
+
+abstract Hash<T>(Dynamic)
+{
+	public inline function keyValueIterator():KeyValueIterator<String, T>
+	{
+		var keys = Reflect.fields(this).iterator();
+		return
+		{
+			hasNext: keys.hasNext,
+			next: ()->
+			{
+				final key = keys.next();
+				return { key:key, value:Reflect.field(this, key)};
+			}
+		};
+	}
+}
+
+typedef AseAtlasHash = AseAtlasBase<Hash<AseAtlasFrame>>;
+typedef TexturePackerAtlasHash = AtlasBase<Hash<TexturePackerAtlasFrame>>;
+
+typedef HashOrArray<T> = flixel.util.typeLimit.OneOfTwo<Hash<T>, Array<T>>;
+
+typedef AseAtlas = AseAtlasBase<HashOrArray<AseAtlasFrame>>;
+typedef TexturePackerAtlas = AtlasBase<HashOrArray<TexturePackerAtlasFrame>>;
+
+typedef AseAtlasMeta =
+{
+	var app:String;
+	var version:String;
+	var image:String;
+	var format:String;
+	var size:AtlasSize;
+	var scale:String;
+	var frameTags:Array<AseAtlasTag>;
+	var slices:Array<AseAtlasSlice>;
+	var layers:Array<AseAtlasLayer>;
+}
+
+typedef AseAtlasTag = { name: String, from:Int, to:Int, direction:String };
+
+typedef AseAtlasSlice =
+{
+	var name:String;
+	var color:String;
+	
+	/**
+	 * Info of at what frames the slice changes size
+	 */
+	var keys: Array<AseAtlasSliceKey>;
 }
 
-typedef TexturePackerFrameRect =
+typedef AseAtlasLayer =
 {
-	x:Float,
-	y:Float,
-	w:Float,
-	h:Float
+	var name:String;
+	
+	/**
+	 * The name of the parent layer
+	 */
+	@:optional var group:String;
+	
+	/**
+	 * Ranges from 0 to 255
+	 */
+	@:optional var opacity:Int;
+	
+	/**
+	 * 
+	 */
+	@:optional var blendMode:AseBlendMode;
+}
+
+enum abstract AseBlendMode(String)
+{
+	var NORMAL = "normal";
+	
+	var DARKEN = "darken";
+	var MULTIPLY = "multiply";
+	var COLOR_BURN = "color_burn";
+	
+	var LIGHTEN = "lighten";
+	var SCREEN = "screen";
+	var COLOR_DODGE = "color_dodge";
+	var ADDITION = "addition";
+	
+	var OVERLAY = "overlay";
+	var SOFT_LIGHT = "soft_light";
+	var HARD_LIGHT = "hard_light";
+	
+	var DIFFERENCE = "difference";
+	var EXCLUSION = "exclusion";
+	var SUBTRACT = "subtract";
+	var DIVIDE = "divide";
+	
+	var HSL_HUE = "hsl_hue";
+	var HSL_SATURATION = "hsl_saturation";
+	var HSL_COLOR = "hsl_color";
+	var HSL_LUMINOSITY = "hsl_luminosity";
+}
+
+typedef AseAtlasSliceKey =
+{
+	/**
+	 * The frame that the slice changes size
+	 */
+	var frame:Int;
+	/**
+	 * The size of the slice at this frame
+	 */
+	var bounds:AtlasRect;
 };
 
-typedef TexturePackerFrameData =
+/**
+ * Size struct use for atlas json parsing, { w:Float, h:Float }
+ */
+typedef AtlasSize = { w:Float, h:Float };
```

Comment:
Aren't these likely to be just type `Int`? I'm currently just casting them to ints, but it got me wondering if floats even make sense here since we are referring to pixel images -- what would at atlas with size `10.3x20.25` be, for example.

Or is this just a parsing safety thing since calling it a `Float` will happily parse it either way? Just mulling this over as I'm implementing stuff in my project.

Reply:
since were dealing in pixels, int is likely the better option. I think I tend to write float by habit